resolver.go

```diff
@@ -501,6 +544,54 @@ func (r *resolver) forwardQueryStart(w dns.ResponseWriter, msg *dns.Msg, queryID
 	return true
 }
 
+func (r *resolver) addQueryToGC(w dns.ResponseWriter, msg *dns.Msg) {
+	var (
+		cc clientConn
+		ok bool
+	)
+	proto := w.LocalAddr().Network()
+	if proto != "udp" {
+		return
+	}
+
+	r.queryLock.Lock()
+	if cc, ok = r.client[msg.Id]; !ok {
+		return
```

Comment:
forgot to unlock.

You could remove the var () block at the top and change the following to:
```
r.queryLock.Lock()
cc, ok := r.client[msg.Id]
r.queryLock.Unlock()
if !ok {
    return
}
```